src/common/sendpacket.c

```diff
@@ -1333,3 +1381,159 @@ sendpacket_abort(sendpacket_t *sp)
+	struct xsk_ring_prod *txr;
+	int ret;
+	xsk = (struct xsk_socket_info*)safe_malloc(sizeof(struct xsk_socket_info));
+	xsk->umem = umem;
+	ret = xsk_socket__create(&xsk->xsk, device, queue_id, umem->umem, rxr, &xsk->tx, cfg);
+	if (ret){
+        return NULL;
+    }
+
+	xsk->app_stats.rx_empty_polls = 0;
+	xsk->app_stats.fill_fail_polls = 0;
+	xsk->app_stats.copy_tx_sendtos = 0;
+	xsk->app_stats.tx_wakeup_sendtos = 0;
+	xsk->app_stats.opt_polls = 0;
+	xsk->app_stats.prev_rx_empty_polls = 0;
+	xsk->app_stats.prev_fill_fail_polls = 0;
+	xsk->app_stats.prev_copy_tx_sendtos = 0;
+	xsk->app_stats.prev_tx_wakeup_sendtos = 0;
+	xsk->app_stats.prev_opt_polls = 0;
+
+	return xsk;
+}
+
+static sendpacket_t * sendpacket_open_xdp(const char *device, char *errbuf){
+    sendpacket_t *sp;
+
+    assert(device);
+    assert(errbuf);
+
+    int nb_of_frames = 131072; // 2^17
+    int frame_size = 4096;
+    int nb_of_completion_queue_desc = 4096;
+    int nb_of_fill_queue_desc = 4096;
+    struct xsk_umem_info* umem_info = create_umem_area(nb_of_frames, frame_size, nb_of_completion_queue_desc, nb_of_fill_queue_desc);
+    if(umem_info == NULL){
+        return NULL;
+    }
+
+    int nb_of_tx_queue_desc = 2048;
+    int nb_of_rx_queue_desc = 2048;
+    u_int32_t queue_id = 0;
+    struct xsk_socket_info* xsk_info = create_xsk_socket(umem_info, nb_of_tx_queue_desc, nb_of_rx_queue_desc, device, queue_id, errbuf);
+    if(xsk_info == NULL){
+        return NULL;
+    }
+
+    sp = (sendpacket_t *)safe_malloc(sizeof(sendpacket_t));
+    strlcpy(sp->device, device, sizeof(sp->device));
+    sp->handle.fd = xsk_info->xsk->fd;
+    sp->handle_type = SP_TYPE_LIBXDP;
+    sp->xsk_info = xsk_info;
+    sp->umem_info = umem_info;
+    sp->frame_size = frame_size;
+    return sp;
+}
+
+struct xsk_umem_info* create_umem_area(int nb_of_frames, int frame_size, int nb_of_completion_queue_descs, int nb_of_fill_queue_descs){
+    int umem_size = nb_of_frames * frame_size;
+    struct xsk_umem_info *umem;
+    void* umem_area = NULL;
+    struct xsk_umem_config cfg = {
+    /* We recommend that you set the fill ring size >= HW RX ring size +
+        * AF_XDP RX ring size. Make sure you fill up the fill ring
+        * with buffers at regular intervals, and you will with this setting
+        * avoid allocation failures in the driver. These are usually quite
+        * expensive since drivers have not been written to assume that
+        * allocation failures are common. For regular sockets, kernel
+        * allocated memory is used that only runs out in OOM situations
+        * that should be rare.
+        */
+    .fill_size = nb_of_fill_queue_descs * 2,
+    .comp_size = nb_of_completion_queue_descs,
+    .frame_size = frame_size,
+    .frame_headroom = 0,
+    .flags = XDP_UMEM_UNALIGNED_CHUNK_FLAG
+    };
+    umem = (struct xsk_umem_info*)safe_malloc(sizeof(struct xsk_umem_info));
+	// if (posix_memalign(&umem_area, getpagesize(), /* PAGE_SIZE aligned */
```

Comment:
this commented part will be removed, I also tried this solution for testing purposes it works correctly, I don't really now which to use posix_memalign or mmap

Reply:
calloc/mmap/posix_memalign all fine. If the memory not aligned the libxdp will gives back EINVAL error. posix_memalign might be better, if later Windows get AF_XDP support for example, posix_memalign will behave same on every platform.

Reply:
Understood, I will use posix_memalign instead of mmap